File: Bitwise/Medium/LET_no29_Divide Two Integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        a, b = abs(dividend), abs(divisor)
        res = 0
        for i in range(31, -1, -1):
            if a >= b << i:
                res += 1 << i
                a -= b << i
        
        if not ((dividend > 0) == (divisor > 0)):
            res = -res
            
        if res < -(2 ** 31):
            return -(2 ** 31)
        if res > 2 ** 31 - 1:
            return 2 ** 31 - 1
        return res

File: Bitwise/Medium/LET_no29_Divide Two Integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == INT_MIN && divisor == -1)
            return INT_MAX;
        
        unsigned int a = abs(dividend), b = abs(divisor), res = 0;
        for (int i = 31; i >= 0; i--) {
            if ((a >> i) >= b) {
                res += 1 << i;
                a -= b << i;
            }
        }

        return (dividend > 0) == (divisor > 0) ? res : -res;
    }
};